corehq/util/couch_doc_processor.py

```diff
@@ -1,16 +1,39 @@
-import hashlib
 import weakref
 from abc import ABCMeta, abstractproperty
 from datetime import datetime
 
 import six
 from couchdbkit import ResourceNotFound
 
-from corehq.util.couch_helpers import paginate_view
-from corehq.util.pagination import PaginationEventHandler, DelegatingPaginationEventHandler
+from corehq.util.couch_helpers import PaginatedViewArgsProvider
+from corehq.util.pagination import PaginationEventHandler, ResumableFunctionIterator, TooManyRetries
 
 
-class ResumableDocsByTypeIterator(object):
+class ResumableDocsByTypeArgsProvider(PaginatedViewArgsProvider):
+    def __init__(self, initial_view_kwargs, doc_types):
+        super(ResumableDocsByTypeArgsProvider, self).__init__(initial_view_kwargs)
+        self.doc_types = doc_types
+
+    def get_next_args(self, result, *last_args, **last_view_kwargs):
+        try:
+            return super(ResumableDocsByTypeArgsProvider, self).get_next_args(result, *last_args, **last_view_kwargs)
+        except StopIteration:
+            last_doc_type = last_view_kwargs["startkey"][0]
+            # skip doc types already processed
+            index = self.doc_types.index(last_doc_type) + 1
+            self.doc_types = self.doc_types[index:]
+            try:
+                next_doc_type = self.doc_types[0]
+            except IndexError:
+                raise StopIteration
+            last_view_kwargs.pop('skip', None)
+            last_view_kwargs.pop("startkey_docid", None)
+            last_view_kwargs['startkey'] = [next_doc_type]
+            last_view_kwargs['endkey'] = [next_doc_type, {}]
+        return last_args, last_view_kwargs
+
+
+def ResumableDocsByTypeIterator(db, doc_types, iteration_key, chunk_size=100, view_event_handler=None):
```

Comment:
What do you think of renaming this to `resumable_docs_by_type_iterator` now that it's a function instead of a class? I was a little confused when reading the code until I realized that it had been changed to a function.